### ISSUE TITLE
ILLinker NRE on Windows

### ISSUE THREAD

**Issue by radical** [dotnet/runtime#69536]
[Build](https://dev.azure.com/dnceng/public/_build/results?buildId=1778207&view=logs&j=a76e47fe-7826-5544-e6de-c2f449cb29d2&t=b4b733d9-5f79-5962-29cc-2464cb6dc677):

```
ILLink : error IL1012: IL Trimmer has encountered an unexpected error. Please report the issue at https://github.com/dotnet/linker/issues [D:\a\_work\1\s\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
##[error]ILLink(0,0): error IL1012: (NETCORE_ENGINEERING_TELEMETRY=Build) IL Trimmer has encountered an unexpected error. Please report the issue at https://github.com/dotnet/linker/issues
  Fatal error in IL Linker
  Unhandled exception. Mono.Linker.InternalErrorException: Step 'OutputStep' failed when processing assembly 'System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e'.
   ---> Mono.Linker.LinkerFatalErrorException: ILLink: error IL1011: Failed to write 'D:\a\_work\1\s\artifacts\obj\coreclr\System.Private.CoreLib\x86\Debug\System.Private.CoreLib.dll'.
   ---> System.NullReferenceException: Object reference not set to an instance of an object.
     at Mono.Cecil.PE.ByteBufferEqualityComparer.Equals(ByteBuffer x, ByteBuffer y)
     at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
     at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
     at Mono.Cecil.Metadata.BlobHeapBuffer.GetBlobIndex(ByteBuffer blob)
     at Mono.Cecil.MetadataBuilder.GetBlobIndex(ByteBuffer blob)
     at Mono.Cecil.MetadataBuilder.GetBlobIndex(Byte[] blob)
     at Mono.Cecil.MetadataBuilder.AddCustomDebugInformations(ICustomDebugInformationProvider provider)
     at Mono.Cecil.MetadataBuilder.AddMethodDebugInformation(MethodDebugInformation method_info)
     at Mono.Cecil.Cil.PortablePdbWriter.Write(MethodDebugInformation info)
     at Mono.Cecil.Cil.CodeWriter.WriteResolvedMethodBody(MethodDefinition method)
     at Mono.Cecil.Cil.CodeWriter.WriteMethodBody(MethodDefinition method)
     at Mono.Cecil.MetadataBuilder.AddMethod(MethodDefinition method)
     at Mono.Cecil.MetadataBuilder.AddMethods(TypeDefinition type)
     at Mono.Cecil.MetadataBuilder.AddType(TypeDefinition type)
     at Mono.Cecil.MetadataBuilder.AddTypes()
     at Mono.Cecil.MetadataBuilder.BuildTypes()
     at Mono.Cecil.MetadataBuilder.BuildModule()
     at Mono.Cecil.MetadataBuilder.BuildMetadata()
     at Mono.Cecil.ModuleWriter.<>c.<BuildMetadata>b__2_0(MetadataBuilder builder, MetadataReader _)
     at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TItem item, Func`3 read)
     at Mono.Cecil.ModuleWriter.BuildMetadata(ModuleDefinition module, MetadataBuilder metadata)
     at Mono.Cecil.ModuleWriter.Write(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters)
     at Mono.Cecil.ModuleWriter.WriteModule(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters)
     at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters)
     at Mono.Cecil.AssemblyDefinition.Write(String fileName, WriterParameters parameters)
     at Mono.Linker.Steps.OutputStep.WriteAssembly(AssemblyDefinition assembly, String directory, WriterParameters writerParameters)
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.OutputStep.WriteAssembly(AssemblyDefinition assembly, String directory, WriterParameters writerParameters)
     at Mono.Linker.Steps.OutputStep.WriteAssembly(AssemblyDefinition assembly, String directory)
     at Mono.Linker.Steps.OutputStep.OutputAssembly(AssemblyDefinition assembly)
     at Mono.Linker.Steps.OutputStep.ProcessAssembly(AssemblyDefinition assembly)
     at Mono.Linker.Steps.BaseStep.Process(LinkContext context)
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.BaseStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger, Boolean throwOnFatalLinkerException)
     at Mono.Linker.Driver.Main(String[] args)
```

|Build|Kind|Start Time|
|---|---|---|
[1813081](https://dev.azure.com/dnceng/public/_build/results?buildId=1813081)|Rolling|2022-08-06|
[1814339](https://dev.azure.com/dnceng/public/_build/results?buildId=1814339)|Rolling|2022-08-06|

**Comment by radical** [comment]
cc @eerhardt

**Comment by eerhardt** [comment]
@radical - has this happened on any other build than your PR? Does it repro in `main`? If yes - do you have repro steps on what kind of build, and on what platform, this repros on?

**Comment by radical** [comment]
No, I have seen it only this one time.

**Comment by eerhardt** [comment]
> No, I have seen it only this one time.

Are you sure the failure isn't caused by your changes? Can you repro it yourself with your changes?

**Comment by radical** [comment]
It passed on the second attempt by azdo for the same commit of my branch. My [PR](https://github.com/dotnet/runtime/pull/68696/files#diff-1a2c2a3ad9fbfb269d653ae3dc750539f97d78bb88a809e9dd08fcb04c033215) has very few non-wasm specific changes, and this is not for a wasm run.

**Comment by eerhardt** [comment]
Hmm, that's odd for a NRE in the linker to be non-deterministic. @agocke @vitek-karas @sbomer - thoughts?

**Comment by AaronRobinsonMSFT** [comment]
Occurred on https://github.com/dotnet/runtime/pull/69619

https://dev.azure.com/dnceng/public/_build/results?buildId=1782592&view=logs&j=81a0f8c2-d4db-561f-9b8e-3f045280d7fe&t=2a104984-c43b-5b93-1361-6bd8a1e96b04&l=960

```

     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.OutputStep.WriteAssembly(AssemblyDefinition assembly, String directory, WriterParameters writerParameters)
     at Mono.Linker.Steps.OutputStep.WriteAssembly(AssemblyDefinition assembly, String directory)
     at Mono.Linker.Steps.OutputStep.OutputAssembly(AssemblyDefinition assembly)
     at Mono.Linker.Steps.OutputStep.ProcessAssembly(AssemblyDefinition assembly)
     at Mono.Linker.Steps.BaseStep.Process(LinkContext context)
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.BaseStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger, Boolean throwOnFatalLinkerException)
     at Mono.Linker.Driver.Main(String[] args)
```

**Comment by MichalStrehovsky** [comment]
If it's indeed non-deterministic, it's probably an issue in the .NET runtime running the tool.

We started to see various mystery bugs in managed tools that run as part of the build ever since switching to the .NET 7 SDK/runtime.

Here's a similar NRE in crossgen2: #69454. The theory is that these are all related, possibly fixed with #68443 and will be fixed once we update to Preview 5 (https://github.com/dotnet/runtime/issues/69593#issuecomment-1132950282).

**Comment by radical** [comment]
Another NRE, though it doesn't look the same. But I'll leave it up to others to decide whether to split it off into a new issue or not. This is on p5 rolling [build](https://dev.azure.com/dnceng/public/_build/results?buildId=1789304&view=logs&j=7768facb-0c9d-5dd5-0fa0-b8bf2c83581b&t=2fe22ea4-f09f-5b78-fe5b-b950e6c53cf3):

```
/_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs(112,17): error IL1005: System.IO.File.Exists(String): Error processing method 'System.IO.FileSystem.FileExists(ReadOnlySpan<Char>)' in assembly 'System.Private.CoreLib.dll'. [D:\a\_work\1\s\src\libraries\Microsoft.Extensions.Logging.Abstractions\tests\Microsoft.Extensions.Logging.Generators.Tests\Microsoft.Extensions.Logging.Generators.Roslyn3.11.Tests.csproj]
##[error]/_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs(112,17): error IL1005: (NETCORE_ENGINEERING_TELEMETRY=Build) System.IO.File.Exists(String): Error processing method 'System.IO.FileSystem.FileExists(ReadOnlySpan<Char>)' in assembly 'System.Private.CoreLib.dll'.
  Mono.Linker.LinkerFatalErrorException: /_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs(112,17): error IL1005: System.IO.File.Exists(String): Error processing method 'System.IO.FileSystem.FileExists(ReadOnlySpan<Char>)' in assembly 'System.Private.CoreLib.dll'.
   ---> System.NullReferenceException: Object reference not set to an instance of an object.
     at Mono.Linker.Steps.MarkStep.IsMethodNeededByInstantiatedTypeDueToPreservedScope(MethodDefinition method)
     at Mono.Linker.Steps.MarkStep.GetRequiredMethodsForInstantiatedType(TypeDefinition type)+MoveNext()
     at Mono.Linker.Steps.MarkStep.MarkRequirementsForInstantiatedTypes(TypeDefinition type)
     at Mono.Linker.Steps.MarkStep.MarkType(TypeReference reference, DependencyInfo reason, Nullable`1 origin)
     at Mono.Linker.Steps.MarkStep.MarkMethodBody(MethodBody body)
     at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method, DependencyInfo& reason, MessageOrigin& origin)
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
     at Mono.Linker.Steps.MarkStep.Process()
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger, Boolean throwOnFatalLinkerException)
```

`System.IO.File.Exists(String): Error processing method 'System.IO.FileSystem.FileExists(ReadOnlySpan<Char>)' in assembly 'System.Private.CoreLib.dll'`

The linker was updated recently with https://github.com/dotnet/runtime/pull/69644 .

cc @vitek-karas @agocke @marek-safar

**Comment by agocke** [comment]
Hmm, that method body hasn't been modified recently. It also doesn't look like it happened in the next build. Seems unlikely to be a linker issue.

**Comment by buyaa-n** [comment]
Another failure on https://dev.azure.com/dnceng/public/_build/results?buildId=1797953&view=logs&j=55f4943c-f76d-585b-7250-deab324f0a54&t=18a666a9-d033-504f-063f-baac7386b04a
```log
ILLink : error IL1005: System.Runtime.InteropServices.SafeHandle.ReleaseHandle(): Error processing method 'Microsoft.Win32.SafeHandles.SafeFileHandle.ReleaseHandle()' in assembly 'System.Private.CoreLib.dll'. [D:\a\_work\1\s\src\libraries\Microsoft.Extensions.Configuration\tests\FunctionalTests\Microsoft.Extensions.Configuration.Functional.Tests.csproj]
##[error]ILLink(0,0): error IL1005: (NETCORE_ENGINEERING_TELEMETRY=Build) System.Runtime.InteropServices.SafeHandle.ReleaseHandle(): Error processing method 'Microsoft.Win32.SafeHandles.SafeFileHandle.ReleaseHandle()' in assembly 'System.Private.CoreLib.dll'.
  Mono.Linker.LinkerFatalErrorException: ILLink: error IL1005: System.Runtime.InteropServices.SafeHandle.ReleaseHandle(): Error processing method 'Microsoft.Win32.SafeHandles.SafeFileHandle.ReleaseHandle()' in assembly 'System.Private.CoreLib.dll'.
   ---> System.NullReferenceException: Object reference not set to an instance of an object.
     at Mono.Linker.Steps.MarkStep.MarkBaseMethods(MethodDefinition method)
     at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method, DependencyInfo& reason, MessageOrigin& origin)
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
     at Mono.Linker.Steps.MarkStep.Process()
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger, Boolean throwOnFatalLinkerException)
D:\a\_work\1\s\.dotnet\sdk\7.0.100-preview.3.22179.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ILLink.targets(108,5): error NETSDK1144: Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false. [D:\a\_work\1\s\src\libraries\Microsoft.Extensions.Configuration\tests\FunctionalTests\Microsoft.Extensions.Configuration.Functional.Tests.csproj]
##[error].dotnet\sdk\7.0.100-preview.3.22179.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ILLink.targets(108,5): error NETSDK1144: (NETCORE_ENGINEERING_TELEMETRY=Build) Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false.
```

**Comment by radical** [comment]
Another NRE, on a [rolling build](https://dev.azure.com/dnceng/public/_build/results?buildId=1802304&view=logs&j=55f4943c-f76d-585b-7250-deab324f0a54&t=18a666a9-d033-504f-063f-baac7386b04a):

```
  Mono.Linker.LinkerFatalErrorException: /_/src/libraries/System.Private.CoreLib/src/System/Object.cs(42,13): error IL1005: System.Object.ToString(): Error processing method 'Microsoft.Diagnostics.Runtime.ClrAppDomain.ToString()' in assembly 'Microsoft.Diagnostics.Runtime.dll'.
   ---> System.NullReferenceException: Object reference not set to an instance of an object.
     at Mono.Collections.Generic.Collection`1.get_Item(Int32 index)
     at Mono.Cecil.MetadataReader.ReadScope(MethodDefinition method)
     at Mono.Cecil.Cil.PortablePdbReader.ReadScope(MethodDebugInformation method_info)
     at Mono.Cecil.Cil.PortablePdbReader.Read(MethodDefinition method)
     at Mono.Cecil.Cil.CodeReader.ReadMethodBody()
     at Mono.Cecil.Cil.CodeReader.ReadMethodBody(MethodDefinition method)
     at Mono.Cecil.MetadataReader.ReadMethodBody(MethodDefinition method)
     at Mono.Cecil.MethodDefinition.<>c.<get_Body>b__41_0(MethodDefinition method, MetadataReader reader)
     at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TRet& variable, TItem item, Func`3 read)
     at Mono.Cecil.MethodDefinition.get_Body()
     at Mono.Linker.Steps.UnreachableBlocksOptimizer.ProcessMethod(MethodDefinition method)
     at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method, DependencyInfo& reason, MessageOrigin& origin)
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
     at Mono.Linker.Steps.MarkStep.Process()
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger, Boolean throwOnFatalLinkerException)
```

**Comment by lewing** [comment]
Do we have a working theory for what is causing these?

**Comment by radical** [comment]
The other recent linker exceptions - https://github.com/dotnet/linker/issues/2789 , and https://github.com/dotnet/runtime/issues/70031 .

**Comment by eerhardt** [comment]
Or a way to get a trace dump so someone can investigate? Has anyone been able to repro this outside of CI?

**Comment by vitek-karas** [comment]
As far as I know we haven't seen this outside of CI - and it's not very common in CI either (less than once a day). The current theory is that it's related to https://github.com/dotnet/runtime/issues/68443.
For example here is a similar failure in crossgen2 (similar as in relatively random NRE): https://github.com/dotnet/runtime/issues/69192

Problem is that crossgen2 and linker both run on the runtime which comes with the SDK and not with the runtime which is live built. So even though the possible cause of this has been fixed for a while, it can still show up for the tools.

As for the crash dump - we have discussed this a little, but I'll continue that discussion today. Right now I don't think we have a way to get a crash dump of these failures from the CI. And even locally it will be hard - it's an exception which is handled by the process, so one would have to run it under debugger to catch this.

**Comment by lewing** [comment]
So I guess we're waiting on https://github.com/dotnet/runtime/pull/70117 which @ViktorHofer doesn't want to land until p5 is officially released?

**Comment by maraf** [comment]
Another one on [#20220603.2](https://dev.azure.com/dnceng/public/_build/results?buildId=1804665&view=logs&j=7768facb-0c9d-5dd5-0fa0-b8bf2c83581b&t=2fe22ea4-f09f-5b78-fe5b-b950e6c53cf3)

```
Mono.Linker.LinkerFatalErrorException: /_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs(112,17): error IL1005: System.IO.File.Exists(String): Error processing method 'System.IO.FileSystem.FileExists(ReadOnlySpan<Char>)' in assembly 'System.Private.CoreLib.dll'.
   ---> System.NullReferenceException: Object reference not set to an instance of an object.
     at Mono.Linker.Steps.MarkStep.IsMethodNeededByInstantiatedTypeDueToPreservedScope(MethodDefinition method)
     at Mono.Linker.Steps.MarkStep.GetRequiredMethodsForInstantiatedType(TypeDefinition type)+MoveNext()
     at Mono.Linker.Steps.MarkStep.MarkRequirementsForInstantiatedTypes(TypeDefinition type)
     at Mono.Linker.Steps.MarkStep.MarkType(TypeReference reference, DependencyInfo reason, Nullable`1 origin)
     at Mono.Linker.Steps.MarkStep.MarkMethodBody(MethodBody body)
     at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method, DependencyInfo& reason, MessageOrigin& origin)
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
     at Mono.Linker.Steps.MarkStep.Process()
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger, Boolean throwOnFatalLinkerException)

```

**Comment by mangod9** [comment]
This doesnt appear to be the potential GC issue. @radical, Is there a dump available to look at?

**Comment by radical** [comment]
> This doesnt appear to be the potential GC issue. @radical, Is there a dump available to look at?

I don't see one. Maybe @vitek-karas has an answer (ref: https://github.com/dotnet/runtime/issues/69536#issuecomment-1146267772).

**Comment by agocke** [comment]
AFAIK this is only happening in builds, and we don't collect crash dumps on build machines. However, we should change the linker to stop swallowing exceptions, https://github.com/dotnet/linker/issues/2824

**Comment by agocke** [comment]
If this was a GC hole, as we suspected, then Preview 5 is merged and this shouldn't happen anymore. Closing until we see new faliures.
